### ISSUE TITLE
GO15VENDOREXPERIMENT

### ISSUE THREAD

**Issue by runningmaster** [kisielk/errcheck#75]
Hi, is there any plans for support new env variable GO15VENDOREXPERIMENT for go1.5 in errcheck? errcheck can't recognize a package if it was placed in $GOPATH/src/vendor folder.

See also for details https://groups.google.com/forum/#!msg/golang-dev/74zjMON9glU/4lWCRDCRZg0J and https://groups.google.com/d/msg/golang-nuts/nnud1w8N3Ik/hPZx32v7DbQJ

Thank you.

**Comment by kisielk** [comment]
Definitely would be good to have. I don't have a ton of time to work on it at the moment, so contributions would be most welcome, otherwise I'm not likely to get to it until it starts to bother me :)

**Comment by runningmaster** [comment]
I'm sorry, probably, the error in this case inside "golang.org/x/tools/go/loader" package, because errcheck gets this error on this line https://github.com/kisielk/errcheck/blob/76bf61ee4096ee72636739d4472186b5de5641db/internal/errcheck/errcheck.go#L101

**Comment by gdm85** [comment]
I also need this feature.

As a preliminary commit, please look at https://github.com/gdm85/gotool/commit/85637c46d0d3e240b4d2e8f253f944114d49235d

Now my question is: is there a more elegant version to match vendor'ed packages rather than modifying matchPackages() itself? e.g. somehow by referencing Go's packages instead? I have been trying to figure out a way, but didn't succeed as all references to vendor'ing seem to lead elsewhere.

(by the way, the whole Go imports logic is a bit hairy already...)

**Comment by kisielk** [comment]
honestly I haven't even tried the vendoring features in go1.5 yet so I'm not really in a position to comment on what works or not

**Comment by gdm85** [comment]
@kisielk vendoring in go1.5 works, there are some edges to smooth but I am positive it will stick in this form (I can't say that I totally like it, but that's another topic).

I am wondering if there is a way to use as fewest lines of code to implement ``vendor`` directory support by using more of the go packages internals; otherwise, I'd just implement it in ``makePackages()`` and submit a PR

**Comment by kisielk** [comment]
By vender directory support do you mean having errcheck be able to find dependancies of the package being checked in a `vendor/` directory?

What is `makePackages()` ?

**Comment by dominikh** [comment]
In my opinion, the right move here is to wait for go/build, go/types and so on to support the vendor experiment. That, in the end, is how all the other tools should support it, too: by using go/build like they've always been.

**Comment by kisielk** [comment]
That's kind of the conclusion I was leaning towards as well.

**Comment by gdm85** [comment]
@kisielk yes. The official specs are here: https://docs.google.com/document/d/1Bz5-UB7g2uPBdOx-rw5t9MxJwkfpx90cqG9AFL0JAYo/edit
Another interesting read: https://medium.com/@freeformz/go-1-5-s-vendor-experiment-fd3e830f52c3

That was a typo, fixed. See also the commit I mentioned, that's the function I changed.

@dominikh that means for me roughly "don't use vendoring or errcheck until the Go Team does that", which I can't accept, hence I was proposing to write a PR to add support now (as many other projects are doing).

@kisielk if I submit a PR would you consider adding the feature **before** it's supported in go/build, go/types etc?

**Comment by kisielk** [comment]
If we were to implement support for vendor, wouldn't it be sufficient to check if there is a `vendor/` directory in the top-level package being checked, and if so, add it to the GOPATH of the build context?

**Comment by kisielk** [comment]
I guess there is still the problem of packages within the vendor directory vendoring packages themselves... that makes the whole situation quite a bit uglier. I'd rather not have to duplicate and maintain the work being done in the main go tree.

**Comment by favadi** [comment]
:+1:, support GO15VENDOREXPERIMENT is a must.

**Comment by kisielk** [comment]
The general solution for this issue is being discussed in golang/go#12278

**Comment by gaillard** [comment]
:+1: for getting this in

**Comment by kisielk** [comment]
Looks like golang/go#12278 has been fixed now so this should work with Go tip. Can someone verify?

**Comment by apriendeau** [comment]
So errcheck fails to build on go 1.6beta1. I dont know if this is related but I am also not diving too deep into it.
```
# golang.org/x/tools/go/loader
/go/src/golang.org/x/tools/go/loader/go16.go:12: undefined: build.IgnoreVendor
```

**Comment by kostya-sh** [comment]
build.AllowVendor (present in Go 1.6beta1) has been changed to
build.IgnoreVendor on tip. This will have to wait till the next beta or rc
release.

**Comment by gaillard** [comment]
Looks like this is still an issue with go1.6beta2.

**Comment by kisielk** [comment]
What's still an issue? It looks like go/loader has been updated for 1.6 with IgnoreVendor. I haven't had an opportunity to try it out yet though.

**Comment by gaillard** [comment]
i just tried to use errcheck from master, is there an update needed in this repo to take advantage of whatever was changed in go/loader?

**Comment by kisielk** [comment]
did you update go/loader as well?

**Comment by gaillard** [comment]
doesn't go get -u update the dependencies as well?

**Comment by gaillard** [comment]
anyone checked this with the 1.6 release?

**Comment by toxeus** [comment]
> anyone checked this with the 1.6 release?
Yes, it works now. Don't forget to recompile `errcheck` and possible dependencies with Go1.6.

**Comment by gaillard** [comment]
Do I need to do anything more than go get -u -f ? Because that doesn't seem to work.

**Comment by toxeus** [comment]
@gaillard I have no idea if `go get -u -f` does the right thing. Assuming you only need to recompile the dependencies without actually updating them then a simple `rm -r $GOPATH/pkg` should do the job.

**Comment by gaillard** [comment]
Seems to work on packages that don't have an eventual dependency on cgo. As soon as it hits one I get things like `error: failed to check packages: could not type check: couldn't load packages due to errors: ...`

**Comment by gaillard** [comment]
https://github.com/golang/go/issues/13526

**Comment by veqryn** [comment]
This is still an issue for me.  See example below, I try to ignore the proto.EncodeVarint function, and errcheck does not ignore.  But if I add `vendor` to the front of the package name, it is successfully ignored.

```
$ bin/errcheck.exe -ignore 'github.com/golang/protobuf/proto:EncodeVarint' api/oib/proto/unsubscriber
src\api\oib\proto\unsubscriber\api.pb.go:206:17 b.EncodeVarint(9<<3 | proto.WireBytes)
src\api\oib\proto\unsubscriber\api.pb.go:211:17 b.EncodeVarint(10<<3 | proto.WireBytes)
src\api\oib\proto\unsubscriber\api.pb.go:216:17 b.EncodeVarint(11<<3 | proto.WireBytes)

$ bin/errcheck.exe -ignore 'vendor/github.com/golang/protobuf/proto:EncodeVarint' api/oib/proto/unsubscriber

$ 
```

**Comment by kisielk** [comment]
@veqryn seems like the `-ignore` flag needs to be more robust with regards to vendor packages. 

I created #102 for tracking that issue. 

I'm going to close this issue since in general the support is working. Any specific problems should get their own issues.

**Comment by veqryn** [comment]
thx

**Comment by ches** [comment]
So, this thread is a bit hard to follow, but if I'm not mistaken the original issue was not actually solved:

> support new env variable GO15VENDOREXPERIMENT for go1.5

The solution is "use Go 1.6 so that errcheck doesn't need to explicitly support the environment variable", is that correct?

If you have a CI build that otherwise worked on Go 1.5 using `vendor/` and setting `GO15VENDOREXPERIMENT`, adding `errcheck` to it then breaks that build. It's unfortunate if I have to  specifically configure around running errcheck on 1.5 builds when many other tools do support the variable, but so be it then.

**Comment by kisielk** [comment]
Well, the fix in `go/build` for `GO15VENDOREXPERIMENT` didn't get shipped until Go 1.6 (https://github.com/golang/go/commit/0c428a56176353d52170f318e998f342b08dacd2) . In order to fix it for this tool we'd have to fork `go/build` locally for use in Go 1.5 builds. That seems like a bit of a nightmare. 

Why don't you just use a version of errcheck built with Go 1.6 for your Go 1.5 CI builds?

**Comment by ches** [comment]
@kisielk I see. I'll give your suggested setup a shot, thanks for the pointers.
